docs: Update migration handling for next version (#2373)
elastisys/compliantkubernetes-apps: migration/README.md

> [!note]
> Only the five latest major and minor releases are kept, if you want to see the migration guides for older versions then you need to check out that version of the repository.

## Writing migration steps for contributions

Notice to developers on writing migration steps:

- Migration steps:
  - are written per minor version and placed in a subdirectory of the migration directory with the name `vX.Y/`,
  - are written to be idempotent and usable no matter which patch version you are upgrading from and to,
  - are documented in this document to be able to run them manually,
  - are divided into prepare and apply steps:
    - Prepare steps:
      - are placed in the `prepare/` directory,
      - may **only** modify the configuration of the environment,
      - may **not** modify the state of the environment,
      - steps are run in order of their names use two digit prefixes.
    - Apply steps:
      - are placed in the `apply/` directory,
      - may **only** modify the state of the environment,
      - may **not** modify the configuration of the environment,

>       - are run in order of their names use two digit prefixes,

Reviewer: small nit
```suggestion
      - are run in order of their names using two digit prefixes,
```

      - are run with the argument `execute` on upgrade and should return 1 on failure and 2 on successful internal rollback,
      - are rerun with the argument `rollback` on execute failure and should return 1 on failure.

> For prepare the init step is given.

Reviewer: Maybe expand this part a bit.
```suggestion
For the prepare part, a init step is always included in each migration which runs the `ck8s init` command.
```

Author: I'm not sure how to rewrite it, because the hope is that we will have more and more persistent migration snippets, so it wont just be init and apply.
I want to remove it, or maybe just state that one should look for the current ones, and give init and apply as examples as a prepare and apply snippet respectively.
:thinking:

Reviewer: Maybe have a link to the respective folders and then provide the init and apply snippets as examples, but I would not say this is blocking in any way

> For apply the bootstrap and the apply steps are given, it is expected that releases upgraded in custom steps are excluded from the apply step.

Reviewer: Could maybe expand this a bit more and clarify how and where you are supposed to exclude certain steps, just to make it easier for people that are writing their first migration script :)

Author: This has been expanded on.

# Upgrade and migration

This directory contains the upgrade and migration steps for each major and minor version, including the documentation with instructions and the snippets for automation of the upgrade process.

> [!note]
> Only the five latest major and minor releases are kept, if you want to see the migration guides for older versions then you need to check out that version of the repository.

## Writing migration steps for contributions

Notice to developers on writing migration steps:

- Migration steps:

>     - are written per major and minor version and placed in a subdirectory of the migration directory with the name `main/`,

Reviewer 2: Nit; feel free to disregard!
```suggestion
    - are written for every major and minor version, and they are placed in a subdirectory of the migration directory with the name `main/`,
```

        - during release this is promoted to the next release series `vX.Y/`,
    - are written to be idempotent and usable no matter which patch version you are upgrading from and to,
    - are documented in `main/README.md` to be able to run them manually,
    - are divided into prepare and apply steps:
        - Prepare steps:
            - are placed in the `prepare/` directory,

>             - may **only** modify the configuration of the environment,
>             - may **not** modify the state of the environment,

Reviewer 2: Not sure if we want to use `an environment` instead of `the environment`.
I'll let you decide.

```suggestion
            - may **only** modify the configuration of an environment,
            - may **not** modify the state of an environment,
```

            - steps are run in order of their names using two digit prefixes.
        - Apply steps:
            - are placed in the `apply/` directory,
            - may **only** modify the state of the environment,
            - may **not** modify the configuration of the environment,
            - are run in order of their names using two digit prefixes,
            - are run with the argument `execute` on upgrade and should return 1 on failure and 2 on successful internal rollback,
            - are rerun with the argument `rollback` on execute failure and should return 1 on failure.

> [Prepare snippets](main/prepare) are supposed to update the environment configuration to the target version, an example snippet is the init step which runs the `ck8s init` command.

Reviewer 2: Nit;
```suggestion
[Prepare snippets](main/prepare) are supposed to update the environment's configuration to the target version, an example snippet is the init step which runs the `ck8s init` command.
```

> [Apply snippets](main/apply) are supposed to update the environment application to the target version, and example sippet is the apply step which run `helmfile upgrade` on all releases that have changed.

Reviewer 3: Typo:

```suggestion
[Apply snippets](main/apply) are supposed to update the environment application to the target version, and example snippet is the apply step which run `helmfile upgrade` on all releases that have changed.
```

> [Apply snippets](main/apply) are supposed to update the environment application to the target version, and example snippet is the apply step which run `helmfile upgrade` on all releases that have changed.

Reviewer 2: Nit;
```suggestion
[Apply snippets](main/apply) are supposed to update the environment's applications to the target versions, and example snippet is the apply step which run `helmfile upgrade` on all releases that have changed.
```

It is expected that releases upgraded in other snippets are excluded from the apply snippet.

> [!tip]
> This can be done by adding label expression of releases into the `skipped*` arrays of the `80-apply.sh` snippet:
>
> ```diff
> --- a/migration/main/apply/80-apply.sh
> +++ b/migration/main/apply/80-apply.sh
> @@ -9,12 +9,15 @@ source "${ROOT}/scripts/migration/lib.sh"
>  # Example: "app!=something"
>  declare -a skipped
>  skipped=(
> +  "app!=something"
>  )
>  declare -a skipped_sc
>  skipped_sc=(
> +  "app!=something"
>  )
>  declare -a skipped_wc
>  skipped_wc=(
> +  "app!=something"
>  )
>
>  run() {
> ```
>
> During runtime these are `AND`:ed together.

Upgrades of components that are dependent on each other should be done within the same snippet to easily manage the upgrade to a working state and to be able to rollback to a working state.

Steps should use the `scripts/migration/lib.sh` which will provide helper functions, see the file for available helper functions.
This script expects the `ROOT` environment variable to be set pointing to the root of the repository.
As with all scripts in this repository `CK8S_CONFIG_PATH` is expected to be set.

> [!important]

> > Since the migration snippets should be idempotent, and usable from and to any patch version, it is important that they only perform actions when needed, especially in the case of destructive actions!

Reviewer 2: Nit;
```suggestion
> Since the migration snippets should be idempotent and usable from and to any patch version, it is important that they only perform actions when needed, especially in the case of destructive actions!
```

>
> Example: if a major upgrade of a component requires that another component is temporarily removed, then it should **only** remove the component if there is a major upgrade in the first place, if there is a minor upgrade then it should be applied without removing the component.

## Preparing migration steps for releases

When releasing new major or minor versions of Welkin Apps the current upgrade and migration directory should be promoted to the one for the release.
To do this run:

```bash
./migration/promote.sh vX.Y
```

This script will create a new directory for the `vX.Y` release series from the `main` directory, and prune old directories to only include the latest five releases.

Once the new directory is created you must manually remove snippets and steps that was specific for the new release, be sure to remove any exclusions from them set in the `80-apply-sh` snippet based on the tip in the previous section.

> General migration steps, such as snippets for upgrading OpenSearch, Prometheus, or Thanos, may be retained as long as they verify beforehand they run that they are needed, and fall back to a generic upgrade steps if not required.

Reviewer 2: I'm a bit confused here and I hope you can clarify it for me :) Once the new release migration directory is created, which snippets and steps should I remove? Everything in `main/`?

Author: No, definitely not all.

The idea is to only remove those snippets that are specific to the version that will never be used again, so we can build up a library of reusable ones that we can iteratively improve, and potentially implement proper rollback steps for.

As an example, [`v0.44/apply/02-migrate-gpu-operator-ns.sh`](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration/v0.44/apply/20-migrate-gpu-operator-ns.sh) is only ever going to make sense once.

While [`v0.44/apply/10-kube-prometheus-stack.sh`](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration/v0.44/apply/10-kube-prometheus-stack.sh) is something we've seen in a few versions, and would be better served with a persistent snippet that would be made a bit more generic.
In its current state it already looks good enough to be kept, as it doesn't even use hardcoded versions.

Perhaps the way to go would be to add some sort of tag to the ones we know should be deleted, so that part of the process is clearer. :thinking:

Reviewer 2: Right, I think I understand now!

Maybe the snippets we want to keep could be moved to a `library/` directory?

Author: Went with a different option to add suffixes for those that should be deleted, as I think that would require special handling in the migration machinery to keep them in a separate directory.

Reviewer 4: So the idea right now is that as long as a migration script that is added is version-generic it should be kept in `main/` and would then be part of all future releases (until it is actively removed)?
My feeling right now is that it might end up being a bit messy if we have like 10-20 generic scripts in `main/` that is added to every release, but only some of them are actually needed in any particular release. But maybe it will not grow to be that big.

I sort of like the idea Olle had that we can move the generic scripts to `library/` after the release is done, then copy them back to `main/` later when they are needed for a release again. One disadvantage I see with that is that we need to consider if unused scripts should be removed from the library at some point.

What are your thoughts on this? Should we reconsider using the library option?

Reviewer 5: This also stumped me a bit, I'm worried that having to remember naming things correctly and removing the correct files etc might be missed or done incorrectly.

Improving the machinery over trusting humans to do the manual process correctly is the way to go IMO.

Author: > So the idea right now is that as long as a migration script that is added is version-generic it should be kept in `main/` and would then be part of all future releases (until it is actively removed)?

Yes, that is my idea. Because then we can reuse and improve migration snippets over time.
As as per how migration snippets work they will only perform upgrade steps when needed, and will degrade to a generic apply if there are regular changes.

> My feeling right now is that it might end up being a bit messy if we have like 10-20 generic scripts in `main/` that is added to every release, but only some of them are actually needed in any particular release. But maybe it will not grow to be that big.

The idea here is for things like opensearch, it has a common flow when you upgrade it, same for kube-prometheus-stack.
Those would make sense to keep, until we don't have opensearch or kube-prometheus-stack, or we've changed their deployment model so the generic apply is enough in all cases.

> I sort of like the idea Olle had that we can move the generic scripts to `library/` after the release is done, then copy them back to `main/` later when they are needed for a release again. One disadvantage I see with that is that we need to consider if unused scripts should be removed from the library at some point.

I would be against this as it would promote bad snippets that always operates, without being required to.
And then it would be the same as just pulling it from an old version.

> What are your thoughts on this? Should we reconsider using the library option?

The only way I would consider that is if we have a linked in library, that are always sourced, but not as a storage place for snippets not in use with the reasoning above.

> This also stumped me a bit, I'm worried that having to remember naming things correctly and removing the correct files etc might be missed or done incorrectly.
> 
> Improving the machinery over trusting humans to do the manual process correctly is the way to go IMO.

How would we make a machinery to evaluate what humans script are required only for one version in specific?
Or what is your idea?

Author: @OlleLarsson @simonklb @viktor-f I'm still waiting on suggestions here.

Reviewer 4: Forgot to answer.
I think that what you say about the generic migration scripts (e.g. for opensearch) makes sense. So I now lean towards your solution being better than the `library/` solution.

Reviewer 5: > > This also stumped me a bit, I'm worried that having to remember naming things correctly and removing the correct files etc might be missed or done incorrectly.
> > Improving the machinery over trusting humans to do the manual process correctly is the way to go IMO.
> 
> How would we make a machinery to evaluate what humans script are required only for one version in specific? Or what is your idea?

My idea is that you would keep the generic snippets that should always run in one place that is kept around for every migration and version specific snippets in a different location so that you don't have to think about the naming or removing some files and keeping some files around.

For example:
migration/include-always/01-opensearch.bash
migration/include-always/02-prometheus.bash
migration/include-always/03-apply.bash
migration/v0.45/01-specific-version-migration.bash

Then it's up to the person introducing the snippets to put them in their correct spot and not the one promoting to the next version.

Author: Yeah, but you have to think of the naming, always, because that is how order is determined, and because of that I see that it is best to keep them in the same directory so it is more intuitive for contributors and reviewers.
Because then what they see is what happens, what is in main will be the next release.

If it is only to identify which snippets are version specific, I rather have a marker in its name because then we can easily identify and automatically prune it when we promote the next version.

```sh
# example
./migration/main/apply/60-prometheus.sh
./migration/main/apply/69-openserch-migrate-snapshots-only-once.sh
./migration/main/apply/70-openserch.sh
./migration/main/apply/80-apply.sh
```

And to add to this, if we must pre-create the migration directory for specific releases when developing specific migrations, then this update won't change anything for what prompted the change in the first place.

Reviewer 5: I think I'm fine with this if we enforce naming so that you have to make the active choice when you introduce a new snippet and then we have automation around what should be left and what should be removed after a promotion. When you promote you should not have to manually pick and choose, it should just be one command and preferably be done as part of cutting a release.

```
./migration/main/apply/60-prometheus-keep.sh
./migration/main/apply/69-openserch-migrate-snapshots-only-once.sh
./migration/main/apply/70-openserch-keep.sh
./migration/main/apply/80-apply-keep.sh
```

Not sure about the naming though.

Reviewer 5: Also, then you can have a linter that enforces the naming scheme so that you are forced to make the choice.

Author: Enforced it with suffixes through pre-commit, chose "all-versions" and "one-version" as "only-once" might make the contributor believe that it will ever only be run once, which isn't the case with patches.

> ```diff
> --- a/migration/main/apply/80-apply-all-versions.sh
> +++ b/migration/main/apply/80-apply-all-versions.sh
> @@ -9,12 +9,15 @@ source "${ROOT}/scripts/migration/lib.sh"
>  # Example: "app!=something"
>  declare -a skipped
>  skipped=(
> +  "app!=something"
>  )
>  declare -a skipped_sc
>  skipped_sc=(
> +  "app!=something"
>  )
>  declare -a skipped_wc
>  skipped_wc=(
> +  "app!=something"
>  )
>
>  run() {
> ```
>
> During runtime these are `AND`:ed together.

Upgrades of components that are dependent on each other should be done within the same snippet to easily manage the upgrade to a working state and to be able to rollback to a working state.

Steps should use the `scripts/migration/lib.sh` which will provide helper functions, see the file for available helper functions.
This script expects the `ROOT` environment variable to be set pointing to the root of the repository.
As with all scripts in this repository `CK8S_CONFIG_PATH` is expected to be set.

> [!important]
> Since the migration snippets should be idempotent and usable from and to any patch version, it is important that they only perform actions when needed, especially in the case of destructive actions!
>

> > Example: if a major upgrade of a component requires that another component is temporarily removed, then it should **only** remove the component if there is a major upgrade in the first place, if there is a minor upgrade then it should be applied without removing the component.

Author: To the component it could be yes, if it is really breaking anything, but that isn't very clear.
I'll reword so `major -> major and minor` and `minor -> patch` to make it clearer, as it is in the context of a patch release we're mainly talking about, so on that we should focus.

We also need them to be idempotent to be able to reuse them between release series, for example to update Grafana, OpenSearch, Prometheus, or Thanos, if not covered by the general apply snippet.
Some snippets will still be specific to a specific version and those snippets should be named with the suffix `-version-specific.sh` to easily identify them.

## Preparing migration steps for releases

When releasing new major or minor versions of Welkin Apps the current upgrade and migration directory should be promoted to the one for the release.
To do this run:

```bash
./migration/promote.sh vX.Y
```

This script will create a new directory for the `vX.Y` release series from the `main` directory, and prune old directories to only include the latest five releases.

Once the new directory is created you must manually remove snippets and steps that was specific for the new release named with the suffix `-version-specific.sh`, be sure to remove any exclusions from them set in the `80-apply.sh` snippet based on the tip in the previous section.

> General migration steps, such as snippets for upgrading Grafana, OpenSearch, Prometheus, or Thanos, may be retained as long as they verify beforehand they run that they are needed, and fall back to a generic upgrade steps if not required.

Reviewer 2: Maybe something like that?
```suggestion
General migration steps, such as snippets for upgrading Grafana, OpenSearch, Prometheus, or Thanos, may be retained as long as they are verified beforehand that they are needed and run when required, and fall back to a generic upgrade steps if not required.
```
